File: Permutations II/Permutations II.cpp
class Solution {
public:
    vector<vector<int>> permuteUnique(vector<int>& nums) {
        vector<vector<int>> result;
        help(nums,result,0);
        return result;
    }
    void help(vector<int>& nums,vector<vector<int>>& result,int begin){
        if(begin==nums.size()){
            result.push_back(nums);
            return;
        }
        unordered_set<int> s;
        for(int i=begin;i<nums.size();++i){
            if(s.find(nums[i])==s.end()){
                swap(nums[begin],nums[i]);
                help(nums,result,begin+1);
                swap(nums[begin],nums[i]);
                s.insert(nums[i]);
            }
        }
        return;
    }
};

File: Solutions/Permutations II/Permutations II.java
class Solution {
    public List<List<Integer>> permuteUnique(int[] nums) {
        List<List<Integer>> res=new ArrayList<>();
        help(res,0,nums);
        return res;
    }
    private void help(List<List<Integer>> res,int begin,int[] nums){
        if(begin==nums.length){
            var temp=new ArrayList<Integer>();
            for(int i=0;i<nums.length;++i)
                temp.add(nums[i]);
            res.add(temp);
            return;
        }
        var s=new HashSet<Integer>();
        for(int i=begin;i<nums.length;++i){
            if(!s.contains(nums[i])){
                swap(nums,begin,i);
                help(res,begin+1,nums);
                swap(nums,begin,i);
                s.add(nums[i]);
            }
        }
        return;
    }
    private void swap(int[] nums,int a,int b){
        int temp=nums[a];
        nums[a]=nums[b];
        nums[b]=temp;
    }
}
